packages/api/main.go

```diff
@@ -114,44 +121,138 @@ func NewGinServer(apiStore *handlers.APIStore, swagger *openapi3.T, port int) *h
+		for idx := range cleanupFns {
+			if cleanup := cleanupFns[idx]; cleanup != nil {
+				wg.Add(1)
+				count++
+				go func() {
+					defer wg.Done()
+					if err := cleanup(); err != nil {
+						exitCode.Add(1)
+						log.Printf("cleanup operation error: %v", err)
+					}
+				}()
+
+				// only run each cleanup once (in case cleanup is called
+				// explicitly.)
+				cleanupFns[idx] = nil
+			}
+		}
+		if count == 0 {
+			log.Println("no cleanup operations")
+			return
+		}
+		log.Printf("running %d cleanup operations", count)
+		wg.Wait() // this doesn't have a timeout
+		log.Printf("%d cleanup operations compleated in %s", count, time.Since(start))
+	}
+	defer cleanup()
+
 	if !env.IsLocal() {
 		shutdown := telemetry.InitOTLPExporter(serviceName, swagger.Info.Version)
-		defer shutdown()
+		cleanupFns = append(cleanupFns, func() error {
+			// shutdown handlers flush buffers upon call and take a context. passing a
+			// specific context here so that all timeout configuration is in one place.
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			return shutdown(ctx)
+		})
 	}
 
 	// Create an instance of our handler which satisfies the generated interface
-	apiStore := handlers.NewAPIStore()
-	defer apiStore.Close()
+	//  (use the outer context rather than the signal handling
+	//   context so it doesn't exit first.)
+	apiStore := handlers.NewAPIStore(ctx)
+	cleanupFns = append(cleanupFns, apiStore.Close)
 
-	s := NewGinServer(apiStore, swagger, *port)
+	// pass the signal context so that handlers know when shutdown is happening.
+	s := NewGinServer(signalCtx, apiStore, swagger, port)
 
-	fmt.Printf("Starting server on port %d\n", *port)
-	// And we serve HTTP until the world ends.
-	err = s.ListenAndServe()
-	if err != nil {
-		log.Printf("server error: %v\n", err)
+	go func() {
+		log.Printf("http service (%d) starting", port)
+
+		// Serve HTTP until shutdown.
+		err := s.ListenAndServe()
+
+		switch {
+		case errors.Is(err, http.ErrServerClosed):
+			log.Printf("http service (%d) shutdown successfully", port)
+		case err != nil:
+			exitCode.Add(2)
+			log.Printf("http service (%d) encountered error: %v", port, err)
+		default:
+			// this probably shouldn't happen...
+			log.Printf("http service (%d) exited without error", port)
+		}
+
+		// cancel the parent context, in case the service
+		// ended outside of shutdown, (we want everything else
+		// to cleanup, and not to get stuck waiting).
+		cancel()
+	}()
+
+	select {
+	case <-signalCtx.Done():
+		// shutdown blocks until all active http handlers have returned.
+		if err := s.Shutdown(ctx); err != nil {
+			exitCode.Add(1)
+			log.Printf("http service (%d) shutdown error: %v", port, err)
+		}
+	case <-ctx.Done():
+		log.Println("http service (%d) shutdown outside of signal", port)
 	}
+
+	// call cleanup explicitly because defers do not run on os.Exit
+	cleanup()
+
+	// Exit, with appropriate code.
+	os.Exit(int(exitCode.Load()))
```

Comment:
LGTM. I have a small question, what's the logic behind the exit code?

Reply:
What do you mean "logic": the exit code is 0 for non-zero "normal" exits, and greater than 1 for others. This is the same as before (exits of > 0 for errors, as a result of calling `log.Fatalf` or `os.Exit(1)`. I'm incrementing this counter from different threads, so the exact number would indicate the number of errors.

The oddity, upon reflection, is that I increment it by 2 on line 230, that should probably just be 1.

Reply:
I thought there's some logic based on `2` and this comment
```
	// Exit, with appropriate code
```

Something like error code 137 is indicates OOM

Reply:
I don't think we can do logic based on the specific return value, but we couldn't already... We could just exit zero or 1, but it seemed interesting to propagate something from the counter?